core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -197,17 +201,25 @@ private CacheEntry getFromContext(InvocationContext ctx, Object key) {
       return cacheEntry;
    }
 
-   private CacheEntry getFromContainer(Object key, int segment, boolean isOwner, boolean writeOperation) {
+   private boolean isPrimaryOwner(boolean isOwner, int segment) {
+      if (distributionManager != null) {
+         return distributionManager.getCacheTopology().getSegmentDistribution(segment).isPrimary();
+      } else {
+         return isOwner;
+      }
+   }
+
+   private CacheEntry getFromContainerForWrite(Object key, int segment, boolean isOwner) {
       if (isOwner) {
-         final InternalCacheEntry ice = innerGetFromContainer(key, segment, writeOperation, false);
+         final InternalCacheEntry ice = innerGetFromContainer(key, segment, false, isPrimaryOwner(isOwner, segment));
          if (trace)
             log.tracef("Retrieved from container %s", ice);
          if (ice == null) {
             return NullCacheEntry.getInstance();
          }
          return ice;
       } else if (isL1Enabled) {
-         final InternalCacheEntry ice = innerGetFromContainer(key, segment, writeOperation, false);
+         final InternalCacheEntry ice = innerGetFromContainer(key, segment, false, isPrimaryOwner(isOwner, segment));
```

Comment:
Could it really be a primary owner if it wasn't an owner?

Reply:
No, but it seemed a little cleaner than adding an && here and hard coding a value in `isPrimaryOwner`. But either way.

Reply:
I was going to change it yeah :)

Reply:
Gotcha, I wasn't look at which branch it was on 👍